### DISCUSSION TITLE
rector conflicts with nikic/php-parser

### DISCUSSION THREAD

**Discussion by staabm** [rectorphp/rector#6958] [General]
I have a project with the following composer.json

```
{
    "name": "staabm/zf-select-strip",
    "license": "MIT",
    "require": {
        "php": "^8.0",
        "nikic/php-parser": "^4.13",
        "rector/rector": "^0.12",
        "phpstan/phpstan": "^1.2",
        "zendframework/zendframework1": "^1.12"
    },
    "require-dev": {
        "ext-pdo": "*",
        "friendsofphp/php-cs-fixer": "3.4.0",
        "phpstan/phpstan-php-parser": "^1.1",
        "phpstan/phpstan-phpunit": "^1.0",
        "phpstan/phpstan-strict-rules": "^1.1",
        "phpunit/phpunit": "^9",
        "symplify/phpstan-extensions": "^10.0"
    },

    "autoload": {
        "psr-4": {
            "staabm\\ZfSelectStrip\\": "src/"
        }
    },
    "autoload-dev": {
        "psr-4": {
            "staabm\\ZfSelectStrip\\Tests\\": "tests/"
        },
        "classmap": [
            "tests/classes/"
        ]
    },
    "scripts": {
        "csfix": [
            "php-cs-fixer fix"
        ],
        "phpstan-dba-baseline": [
            "phpunit",
            "phpstan analyse -c phpstan.neon.dist"
        ],
        "phpstan": [
            "phpstan analyse -c phpstan.neon.dist"
        ],
        "phpunit": [
            "phpunit"
        ]
    },
    "config": {
        "optimize-autoloader": true,
        "sort-packages": true,
        "allow-plugins": {
            "cweagans/composer-patches": true,
            "rector/extension-installer": true
        }
    }
}
```

as soon as I use
``` 
        "nikic/php-parser": "^4.13",
        "rector/rector": "^0.12",
```

in tandem , I get errors when running phpunit

```
PHP Fatal error:  Cannot declare interface PhpParser\Node, because the name is already in use in C:\dvl\Workspace\zf-select-strip\vendor\rector\rector\vendor\nikic\php-parser\lib\PhpParser\Node.php on line 6
```

it seems the problem is, that rector itself includes nikic/php-parser again.

how is the correct way to declare the dependency to rector in a project in which I want to develop new phpstan extensions in tandem with new rectors?

**Comment by staabm** [comment]
said differently: I am using some dev tools which depend on nikic/php-parser:
```
$ composer why nikic/php-parser
phpunit/php-code-coverage  9.2.10   requires  nikic/php-parser (^4.13.0)
sebastian/complexity       2.0.2    requires  nikic/php-parser (^4.7)
sebastian/lines-of-code    1.0.3    requires  nikic/php-parser (^4.6)
symplify/astral            10.0.14  requires  nikic/php-parser (^4.13.2)
```

and even if my project does not declare a direct dependency to `nikic/php-parser`  the one embedded in rector stays in conflict with the one in the project (which gets installed because transitive dependency)

using rector `0.12.13`


-----

the rector embedded php-parser inside vendor/rector/rector/vendor/
![grafik](https://user-images.githubusercontent.com/120441/150521118-3de0406a-e3c3-4df4-9c2f-4987ba2924aa.png)


the global one in the projects vendor/

![grafik](https://user-images.githubusercontent.com/120441/150521159-2fe2af71-55c3-4abd-b65c-076893dadc4a.png)

**Comment by TomasVotruba** [comment]
Thanks for reporting the issue. We're using Rector and packages that use php-parser in many projects and it runs propperly.
This looks like autoloading conflict. Could you reproduce in repository on Github with failing action? I'll investigate, where is gets stuck.

**Reply by staabm** [reply]
please see https://github.com/staabm/zf-select-strip/pull/5 and the phpunit ci job at commit `cdcb39e`

https://github.com/staabm/zf-select-strip/runs/4896396830?check_suite_focus=true

![grafik](https://user-images.githubusercontent.com/120441/150529256-0e86ac9d-d928-49ed-96f8-ec799a9e146f.png)

**Reply by staabm** [reply]
as a workaround on my local machine I commented the require-once calls of `preload.php`

**Reply by samsonasik** [reply]
@staabm could you verify if this PR solve the issue ?

**Reply by staabm** [reply]
maybe rector should declare that it provides `nikic/php-parser`?

https://getcomposer.org/doc/04-schema.md#provide

**Reply by samsonasik** [reply]
The phpunit redeclare issue usually solved by run to directory directly, eg:

```bash
vendor/bin/phpunit tests
```

**Reply by TomasVotruba** [reply]
@samsonasik Agreed, I came to a similar solution.

**Comment by TomasVotruba** [comment]
Here are few notes: https://github.com/staabm/zf-select-strip/pull/8/files

**Comment by Lewiscowles1986** [comment]
~Question if I might.~ So is the https://github.com/rectorphp/rector/tree/main/src/PhpParser actually nikic/php-parser? If So why on earth is it not in the composer.json? 🤯 

Answered myself (see comment)

**Reply by Lewiscowles1986** [reply]
Hmm. Seems like this is a difference between `rectorphp/rector` and `rectorphp/rector-src` ... [see rector-src composer.json](https://github.com/rectorphp/rector-src/blob/main/composer.json)

Question withdrawn, but this seems very strange.

**Reply by TomasVotruba** [reply]
We're using:

* development repository (on PHP 8.1) - just for contributors, https://github.com/rectorphp/rector-src
* and release repository (on PHP 7.2 and scoped) - for yourself, end-used https://github.com/rectorphp/rector

Check post on Rector blog that explains it with examples: https://getrector.org/blog/how-to-bump-minimal-version-without-leaving-anyone-behind

**Comment by dereuromark** [comment]
For us, the only thing that seemd to help was to disable coverage:
https://github.com/cakephp/upgrade/pull/366
since the conflict happens when php-code-coverage loads php-parser.

**Comment by TomasVotruba** [comment]
Locking as stale.
